src/EGraphs/saturation.jl

```diff
@@ -198,49 +261,39 @@ function eqsat_apply!(
   theory::Theory,
   rep::SaturationReport,
   params::SaturationParams,
-  ematch_buffer::OptBuffer{UInt128},
+  ematch_buffer::OptBuffer{UInt64},
 )
   n_matches = 0
-  k = length(ematch_buffer)
-
-  @debug "APPLYING $(count((==)(0xffffffffffffffffffffffffffffffff), ematch_buffer)) matches"
   g.needslock && lock(g.lock)
-  while k > 0
 
+  k = 1
+  while k < length(ematch_buffer)
     if n_matches % CHECK_GOAL_EVERY_N_MATCHES == 0 && params.goal(g)
       @debug "Goal reached"
       rep.reason = :goalreached
       return
     end
 
-    delimiter = ematch_buffer.v[k]
-    @assert delimiter == 0xffffffffffffffffffffffffffffffff
-    n = k - 1
-
-    next_delimiter_idx = 0
-    n_elems = 0
-    for i in n:-1:1
-      n_elems += 1
-      if ematch_buffer.v[i] == 0xffffffffffffffffffffffffffffffff
-        n_elems -= 1
-        next_delimiter_idx = i
-        break
-      end
-    end
-
     n_matches += 1
-    match_info = ematch_buffer.v[next_delimiter_idx + 1]
-    id = v_pair_first(match_info)
-    rule_idx = reinterpret(Int, v_pair_last(match_info))
+
+
+    id = ematch_buffer.v[k]
+    rule_idx = reinterpret(Int, ematch_buffer.v[k + 1])
+    isliteral_bitvec = ematch_buffer.v[k + 2]
     direction = sign(rule_idx)
     rule_idx = abs(rule_idx)
     rule = theory[rule_idx]
 
-    bindings = @view ematch_buffer.v[(next_delimiter_idx + 2):n]
+    bind_start = k + 3
+
+    bind_end = bind_start + length(rule.patvars) - 1
+
+    bindings = @view ematch_buffer.v[bind_start:bind_end]
+
+    res = apply_rule!(bindings, isliteral_bitvec, g, rule, id, direction)
 
-    res = apply_rule!(bindings, g, rule, id, direction)
+    k = bind_end + 1
```

Comment:
This is indeed much cleaner now. I would only like to add `getindex()` for ematch_buffer, so that we do not need to access `ematch_buffer.v` directly.

Reply:
Note that we now have 3x UInt64 instead of 1x UInt128 for the match_info